### PR TITLE
[12.0][ADD] Add delivery service

### PR DESCRIPTION
**Pull request by lmignon**
Forward port of #325 
requires https://github.com/OCA/server-tools/pull/1711

### CODE REVIEW

File: shopinvader_delivery_carrier/tests/test_delivery_service.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        :param pickings: stock.picking recordset
+        :return: bool
+        """
+        # To have them into correct order
+        pickings = pickings.search([("id", "in", pickings.ids)])
+        self.assertEquals(len(data), len(pickings))
+        for current_data, picking in zip(data, pickings):
+            carrier_dict = current_data.get("carrier", {})
+            sale_dict = current_data.get("sale", {})
+            self.assertEquals(current_data.get("delivery_id"), picking.id)
```
Comment (simahawk) [inline shopinvader_delivery_carrier/tests/test_delivery_service.py]: `assertEquals` is sort of deprecated in py3, use `assertEqual`

File: shopinvader_delivery_carrier/tests/test_delivery_service.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                self.assertEquals(
+                    carrier_dict.get("name"), picking.carrier_id.name
+                )
+            else:
+                self.assertFalse(carrier_dict)
+            if picking.sale_id:
+                self.assertEquals(
+                    sale_dict.get("state"), picking.sale_id.state
+                )
+                self.assertAlmostEquals(
```
Comment (simahawk) [inline shopinvader_delivery_carrier/tests/test_delivery_service.py]: same `-s`

File: shopinvader_delivery_carrier/tests/test_delivery_service.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+            partner=self.service.partner, sale=True
+        )
+        picking3 = self._create_picking(
+            partner=self.service.partner, sale=True
+        )
+        picking4 = self._create_picking(
+            partner=self.service.partner, sale=True
+        )
+        pickings = picking1 | picking2 | picking3 | picking4
+        self.assertEquals(picking1.partner_id, self.service.partner)
```
Comment (simahawk) [inline shopinvader_delivery_carrier/tests/test_delivery_service.py]: these assertions seem superfluous as you've just created the pickings w/ this partner

**Review by simahawk — APPROVED** [review]
LG. Minor changes for tests, no blockers.

@lmignon additional consideration for this and other modules: we should include in README some documentation on the usage and the behavior of the rest endpoints in regards of the client (locomotive basically but better if generic). 
Something like: "on your website client you can call this service like $snippet and you have to configure $this and take care of $that." 
This is especially true for important changes like the one that happened on payments "this change requires you update you client template like $this and change $that" etc.
I mean, I know that the client should have its own docs but the functionalities are declared server side as well as the way someone should interact w/ them. IMO we must have some generic client-oriented documentation in each module. What do you think?

**Comment by lmignon** [comment]
> LG. Minor changes for tests, no blockers.
> 
> @lmignon additional consideration for this and other modules: we should include in README some documentation on the usage and the behavior of the rest endpoints in regards of the client (locomotive basically but better if generic).
> Something like: "on your website client you can call this service like $snippet and you have to configure $this and take care of $that."
> This is especially true for important changes like the one that happened on payments "this change requires you update you client template like $this and change $that" etc.
> I mean, I know that the client should have its own docs but the functionalities are declared server side as well as the way someone should interact w/ them. IMO we must have some generic client-oriented documentation in each module. What do you think?

@simahawk I fully agree with you. It's the most important missing part of this project. We must also provide a clear changelog. I would like to start using [towncrier](https://github.com/hawkowl/towncrier)  to manage the changelog. @sbidoul plan to run towncrier in the bumpversion step of "ocabot merge" to generate the changelog https://github.com/OCA/maintainer-tools/pull/432

**Pull request merged**